src/hotspot/share/memory/metaspace.cpp

```diff
@@ -805,29 +806,33 @@ void Metaspace::global_initialize() {
     // Set up compressed class pointer encoding.
     // In CDS=off mode, we give the JVM some leeway to choose a favorable base/shift combination.
     CompressedKlassPointers::initialize((address)rs.base(), rs.size());
+
+    // After narrowKlass encoding scheme is decided: if the encoding base points to class space start,
+    // establish a protection zone.
+    if (CompressedKlassPointers::base() == (address)rs.base()) {
```

Comment:
What happens if this condition is not true? Do we need  to check for encoded values of zero?

Or, is the following true?

```
} else {
    assert((unitx)(CompressedKlassPointers::base()) == 0x0, "must be zero based");
}
```

Is the `else` part covered by your new test case?

Reply:
Its the latter (even though base could be anywhere in front of the Klass range as long as the encoding covers the class space, I never used that since I planned to do this zero-access page for a while now). I added the else assert.